catalog/db_branch_scanner.go

```diff
@@ -35,9 +42,36 @@ func NewDBBranchScanner(tx db.Tx, branchID int64, commitID CommitID, opts *DBSca
 		s.opts.BufferSize = DBScannerDefaultBufferSize
 	}
 	s.buf = make([]*DBScannerEntry, 0, s.opts.BufferSize)
+	commitsWhere, _ := getRelevantCommitsCondition(tx, branchID, commitID)
+	s.commitsWhere = commitsWhere
 	return s
 }
 
+func getRelevantCommitsCondition(tx db.Tx, branchID int64, commitID CommitID) (string, error) {
+	var maxCommitId CommitID
+	var commits []string
+	var commitsWhere string
+	if commitID == UncommittedID || commitID == CommittedID {
+		maxCommitId = MaxCommitID
+	} else {
+		maxCommitId = commitID
+	}
+	sql := "SELECT commit_id::text FROM catalog_commits WHERE branch_id = $1 AND commit_id <= $2 ORDER BY commit_id"
+	err := tx.Select(&commits, sql, branchID, maxCommitId)
+	if err != nil {
+		panic(err)
+	}
+	if commitID == UncommittedID {
+		commits = append(commits, strconv.FormatInt(int64(MaxCommitID), 10))
+	}
+	if len(commits) > MaxCommitsInFilter {
```

Comment:
Isn't this condition backwards?  I'd expect to want to have all the commits if there are *few* commits, otherwise use a range.  As written `commitsWhere` has unbounded length, which is odd.

Reply:
Thanks - you are right.Could have saved me time.   I found it with the debugger :-(
Very impressive

Reply:
About the general comment - F2F

Reply:
Thanks!  This makes a lot more sense.  (Seeing "> max" is usually a red flag.)  But this version (without the fix) could not have been any faster -- it yielded essentially the same SQL  So we must have a repeatable benchmark for this code to know that it is faster -- e.g. it would probably have caught this error.

After flipping the comparison, the my comment about an unbounded query length is no longer relevant of course.